### ISSUE TITLE
apitrace doesn't trace double lock properly, causing visual corruption in traces of affected games (Risen 3)

### ISSUE THREAD

**Issue by aufkrawall** [apitrace/apitrace#697]
See this comment by Gallium Nine developer Axel Davy for reference:
https://github.com/iXit/wine-nine-standalone/issues/40#issuecomment-748605186

> The artifacts are due to an apitrace bug when the trace is made, as shown here:
> 
> ![Screenshot_20201220_140039](https://user-images.githubusercontent.com/2546609/102713939-e86a3180-42cb-11eb-8c69-7af9579da5d4.png)
> 
> apitrace seems to have trouble tracking double lock.

This causes corruption in traces for the game Risen 3. A sample trace can be found here: https://drive.google.com/file/d/1JJnVuUFWgx53DObZYXZT7XfOFmW1jfIj/view

The game itself runs without visual corruption with the AMD D3D9 driver on Windows 10, but when replaying a captured trace with the same driver, there is heavy visual corruption. A recent apitrace appveyor build as of today was used.

**Comment by axeldavy** [comment]
The correct behaviour here would be to have a memcpy of size 6144 with pointer 0x12400200 and one of size 3329 with pointer 0x12401a00. Not two of the latter.

In addition as the data can be filled by two separate threads, we cannot know which unlock is associated to which lock.
Thus I believe the first unlock is noop in the d3d9 implementation, and only the last one does something.
Thus apitrace should record the content written only at the moment of the last unlock.

**Comment by jrfonseca** [comment]
Yes, in https://github.com/apitrace/apitrace/blob/master/wrappers/d3d9trace.py#L62 instead of keeping one map per sub-resource, we should keep a stack of mappings.

But note that double locks are always ambiguous -- as one never knows if the unlocks methods are called in FIFO or LIFO order.

Perhaps an even better thing to do is to, keep a reference count of the number of maps, and whenever there's a nested lock do two things:
- increase the map count
- expand the mapping range to be the _union_ of the previous map and the current map

Alas this is all complicated.  Even though some games hit this now and then, only few do, and it's a hassle to fix.  I don't think I'll have time for the foreseable future.  I'd happily accept patches if anybody wants to give it a go.
